### DISCUSSION TITLE
How do I train a multi-speaker VITS model?

### DISCUSSION THREAD

**Discussion by BoneGoat** [coqui-ai/TTS#879] [General Q&A]
Hi,

I have checked the docs, followed the code, but none of my attempts have been successful.

My understanding is that setting use_speaker_embedding = True would trigger the code path to enable multi-speaker, generate speakers.json, etc. I am feeding the speaker_name of each audio sample using my dataset formatter. I even tried to set num_speakers in the config but that would just crash the training with an unknown parameter error. I'm using the VitsConfig class to configure the experiment.

So, are there any docs I've missed that explains this?

Best regards

**Comment by BoneGoat** [comment]
Using the released model (https://github.com/coqui-ai/TTS/releases/download/v0.2.0/tts_models--en--vctk--vits.zip) as reference I managed to get a little bit further.

There are two use_speaker_embedding in the config, the one I was looking for is: model_args.use_speaker_embedding which should be set to True.

speaker_ids.json (speakers_file in the config) is a mapping between the names given by the dataset formatter and an index: { "Speaker1": 0, "Speaker2": 1 }

I also set num_speakers to the number of speakers in my dataset.

Using these settings the multi-speaker training will start but crash at: https://github.com/coqui-ai/TTS/blob/0592a5805ca7fb877c8fc8df56a4eacac13f2657/TTS/tts/models/vits.py#L669
Because emb_g is an Embedding object and not iterable. Changing the line to: gen_parameters = chain(gen_parameters, self.emb_g.parameters()) will fix that.

The training will still crash though... And I think I've reached my limit here.

```
Traceback (most recent call last):
  File "/workspace/TTS/trainer.py", line 919, in fit
    self._fit()
  File "/workspace/TTS/trainer.py", line 904, in _fit
    self.train_epoch()
  File "/workspace/TTS/trainer.py", line 738, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/workspace/TTS/trainer.py", line 623, in train_step
    outputs, loss_dict_new, step_time = self._optimize(
  File "/workspace/TTS/trainer.py", line 497, in _optimize
    outputs, loss_dict = self._model_train_step(batch, model, criterion, optimizer_idx=optimizer_idx)
  File "/workspace/TTS/trainer.py", line 458, in _model_train_step
    return model.train_step(*input_args)
  File "/workspace/TTS/tts/models/vits.py", line 521, in train_step
    outputs = self.forward(
  File "/workspace/TTS/tts/models/vits.py", line 432, in forward
    o = self.waveform_decoder(z_slice, g=g)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1056, in _call_impl
    return forward_call(*input, **kwargs)
  File "/workspace/TTS/vocoder/models/hifigan_generator.py", line 250, in forward
    o = o + self.cond_layer(g)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1056, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 301, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 297, in _conv_forward
    return F.conv1d(input, weight, bias, self.stride,
TypeError: conv1d() received an invalid combination of arguments - got (NoneType, Parameter, Parameter, tuple, tuple, tuple, int), but expected one of:
 * (Tensor input, Tensor weight, Tensor bias, tuple of ints stride, tuple of ints padding, tuple of ints dilation, int groups)
      didn't match because some of the arguments have invalid types: (NoneType, Parameter, Parameter, tuple, tuple, tuple, int)
 * (Tensor input, Tensor weight, Tensor bias, tuple of ints stride, str padding, tuple of ints dilation, int groups)
      didn't match because some of the arguments have invalid types: (NoneType, Parameter, Parameter, tuple, tuple, tuple, int)
```

**Comment by erogol** [comment]
You can now check the Dev branch. There are recipes for VCTK/

**Comment by tieincred** [comment]
@BoneGoat Hello, 
Please forgive for naïve question as I am need to the tech.
I am trying to train a multi speaker model on vits but I am unable to understand do we have to calculate the d_vectors for speakers using the TTS/bin/compute_embeddings.py file?
Or using the recipe it'll be calculated automatically, because there is nothing mentioned as such in recipe or am I missing something?
